From 407acfeba87b3b58d8e0493f5efff1bb4cee32bd Mon Sep 17 00:00:00 2001
From: Henry Schreiner <HenrySchreinerIII@gmail.com>
Date: Tue, 28 Dec 2021 14:32:29 -0500
Subject: [PATCH] fix: support conda-forge

---
 setup.py | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/setup.py b/setup.py
index a2326287d8..2c44d25d8d 100644
--- a/setup.py
+++ b/setup.py
@@ -146,6 +146,8 @@ def remove_output(*sources):
             "-DBUILD_TESTING=OFF",
             "-DPYBIND11_NOPYTHON=ON",
         ]
+        if "CMAKE_ARGS" in os.environ:
+            cmd = cmd[:5] + os.environ["CMAKE_ARGS"].split() + cmd[5:]
         cmake_opts = dict(cwd=DIR, stdout=sys.stdout, stderr=sys.stderr)
         subprocess.check_call(cmd, **cmake_opts)
         subprocess.check_call(["cmake", "--install", tmpdir], **cmake_opts)
